File: algo/24. Swap Nodes in Pairs.py
import json

class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    def swapPairs(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if head is None:
            return None
        if head.next is None:
            return head

        prevHead = ListNode(0)
        prevHead.next = head
        prevNode = prevHead
        currNode = head

        while currNode is not None:
            if currNode.next is None:
                break
            nextNode = currNode.next
            currNode.next = nextNode.next
            nextNode.next = currNode
            prevNode.next = nextNode
            prevNode = currNode
            currNode = currNode.next

        return prevHead.next


def stringToIntegerList(input):
    return json.loads(input)

def stringToListNode(input):
    # Generate list from the input
    numbers = stringToIntegerList(input)

    # Now convert that list into linked list
    dummyRoot = ListNode(0)
    ptr = dummyRoot
    for number in numbers:
        ptr.next = ListNode(number)
        ptr = ptr.next

    ptr = dummyRoot.next
    return ptr

def listNodeToString(node):
    if not node:
        return "[]"

    result = ""
    while node:
        result += str(node.val) + ", "
        node = node.next
    return "[" + result[:-2] + "]"

def main():
    import sys
    import io
    def readlines():
        for line in io.TextIOWrapper(sys.stdin.buffer, encoding='utf-8'):
            yield line.strip('\n')

    lines = readlines()
    while True:
        try:
            line = next(lines)
            head = stringToListNode(line);
            
            ret = Solution().swapPairs(head)

            out = listNodeToString(ret);
            print(out)
        except StopIteration:
            break

if __name__ == '__main__':
    main()

File: round2/24. Swap Nodes in Pairs.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode SwapPairs(ListNode head) {
        ListNode dummy = new ListNode(-1);
        dummy.next = head;
        ListNode c1 = dummy;
        ListNode c2 = head;
        while (c1 != null && c2 != null && c2.next != null) {
            c1.next = c2.next;
            ListNode temp = c1.next.next;
            c1.next.next = c2;
            c2.next = temp;
            c1 = c1.next.next;
            c2 = c2.next;
        }
        return dummy.next;
    }
}